presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -0,0 +1,160 @@
+import net.jcip.annotations.ThreadSafe;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+@ThreadSafe
+public class MultilevelSplitQueue
+{
+    public static final int[] LEVELS = {0, 1, 10, 60, 300};
+
+    private final List<PriorityQueue<PrioritizedSplitRunner>> levelWaitingSplits;
+    private final AtomicLong[] levelMinPriority;
+
+    private final ReentrantLock lock;
+    private final Condition notEmpty;
+
+    private final List<CounterStat> selectedLevelCounters;
+
+    public MultilevelSplitQueue()
+    {
+        this.levelMinPriority = new AtomicLong[LEVELS.length];
+        this.levelWaitingSplits = new ArrayList<>(LEVELS.length);
+        this.selectedLevelCounters = new ArrayList<>(LEVELS.length);
+
+        for (int i = 0; i < LEVELS.length; i++) {
+            levelMinPriority[i] = new AtomicLong(-1);
+            levelWaitingSplits.add(new PriorityQueue<>());
+            selectedLevelCounters.add(new CounterStat());
+        }
+
+        this.lock = new ReentrantLock();
+        this.notEmpty = lock.newCondition();
+    }
+
+    public void offer(PrioritizedSplitRunner split)
+    {
+        lock.lock();
+        split.setReady();
+        levelWaitingSplits.get(split.getLevel()).offer(split);
+        notEmpty.signal();
+        lock.unlock();
+    }
+
+    public PrioritizedSplitRunner take()
+            throws InterruptedException
+    {
+        PrioritizedSplitRunner result;
+        while (true) {
+            lock.lockInterruptibly();
+            try {
+                while ((result = takeFirstSplit()) == null) {
+                    notEmpty.await();
+                }
+
+                if (result.updateLevelPriority()) {
+                    offer(result);
+                    continue;
+                }
+
+                int selectedLevel = result.getLevel();
+                levelMinPriority[selectedLevel].set(result.getLevelPriority());
+                selectedLevelCounters.get(result.getLevel()).update(1);
+
+                return result;
+            }
+            finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Call only while holding lock.
+     */
+    private PrioritizedSplitRunner takeFirstSplit()
+    {
+        for (int level = 0; level < LEVELS.length; level++) {
+            PrioritizedSplitRunner split = levelWaitingSplits.get(level).poll();
+            if (split != null) {
+                return split;
+            }
+        }
+
+        return null;
+    }
+
+    public void remove(PrioritizedSplitRunner split)
+    {
+        lock.lock();
+        for (int level = 0; level < LEVELS.length; level++) {
+            levelWaitingSplits.get(level).remove(split);
```

Comment:
It is strictly safer to check every list, since the client could have modified the level.